src/error_diffusion.jl

```diff
@@ -1,38 +1,55 @@
-function error_diffusion(
-    img::AbstractMatrix{<:Gray}, stencil::OffsetMatrix; to_linear=false
-)::BitMatrix
+struct ErrorDiffusion{S,RR,CR} <: AbstractDither
+    stencil::S
+    rowrange::RR
+    colrange::CR
+
+    function ErrorDiffusion(stencil, rowrange, colrange)
+        stencil = OffsetMatrix(stencil, rowrange, colrange)
+        return new{typeof(stencil),typeof(rowrange),typeof(colrange)}(
+            stencil, rowrange, colrange
+        )
+    end
+end
+
+function dither(
+    img::AbstractMatrix{T},
+    alg::ErrorDiffusion,
+    cs::ColorScheme;
+    to_linear=false,
+    metric::DifferenceMetric=DE_2000(),
+)::AbstractMatrix{T} where {T<:Color}
     # this function does not yet support OffsetArray
     require_one_based_indexing(img)
 
+    length(cs) >= 2 ||
+        throw(DomainError(steps, "Color scheme for dither needs >= 2 colors."))
+
     # Change from normalized intensities to Float as error will get added!
-    FT = floattype(eltype(img))
+    FT = floattype(T)
     if to_linear
         _img = @. FT(srgb2linear(img))
     else
         _img = FT.(img)
     end
-    stencil = FT.(stencil) # eagerly promote to the same type to make loop run faster
+    stencil = alg.stencil # eagerly promote to the same type to make loop run faster
```

Comment:
It would be nice to keep the `FT`, but I didn't get it to work with color images. It caused errors in this line:
https://github.com/adrhill/DitherPunk.jl/blob/b80ed2312e1efcd3cc7f987886687bb4e7dfd09d/src/error_diffusion.jl#L54

where multiplication between two RGB values is not defined. 

`stencil` is actually just a matrix of `Rational`s summing up to approximately 1, which distribute the error `err` within the neighbourhood defined by them.

Reply:
I see. In this case, I think we only need to extract the storage type information from given `FT`, i.e., `eltype(FT)`:

```suggestion
    stencil = eltype(FT).(alg.stencil) # eagerly promote to the same type to make loop run faster
```

Could you verify if it works?